Added basic gating as TimeSeries method

tjma12 opened the pull request:
This PR adds gating as a `TimeSeries` class method. The method whitens the time series, marks all points above a threshold, clusters them within a time window, and applies inverse Planck windows to zero out the data. The **unwhitened**, gated time series is returned.

An example can be found here (behind LIGO.org auth): https://ldas-jobs.ligo.caltech.edu/~thomas.massinger/notebook/demonstrate-gwpy-gating.html

Whitening arguments can be passed to the underlying whitening used to locate gating points.

pep8speaks commented:
Hello @tjma12! Thanks for submitting the PR.

 - In the file [`gwpy/timeseries/timeseries.py`](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py), following are the PEP8 issues :

> [Line 60:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L60): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 62:73](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L62): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 71:76](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L71): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 87:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L87): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 91:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L91): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 91:90](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L91): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 94:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L94): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1606:53](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1606): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1609:77](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1609): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1611:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1611): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1615:77](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1615): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1621:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1621): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1625:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1625): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1632:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1632): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 1638:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1638): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 1645:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1645): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1648:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1648): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)
> [Line 1651:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1651): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 1660:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1660): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 1671:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1671): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1675:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1675): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1677:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1677): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 1680:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1680): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1683:80](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1683): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 1685:1](https://github.com/gwpy/gwpy/blob/2b105780783f65d8b2774b913525169e19cd18b2/gwpy/timeseries/timeseries.py#L1685): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

alurban commented:
@duncanmmacleod, the build failure reported by Travis seems to be from `pycbc` trying to import `glue.segments`

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -58,6 +58,45 @@ def _fft_length_default(dt):
     return int(max(2, numpy.ceil(2048 * dt.decompose().value)))
 
 
+def _discover_gate_points(ts, thresh, window):
```
Is this method a generic 'points above threshold' detector? If so, can we move it into the `Series` class as an instance method and give it an appropriate name?

tjma12 replied:
It's a "points above threshold" detector with the automatic clustering that you'd want for gating. If I move it into the `Series` class, would you like me to make the clustering optional?

tjma12 replied:
Will replace this with `scipy.signal.find_peaks`

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1606,98 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+            gates = _discover_gate_points(self.whiten(**whiten_kwargs),
+                                          threshold, cluster_window)
+        else:
+            gates = _discover_gate_points(self, threshold, cluster_window)
```
I would modify this to:

```python
        data = self.whiten(**whiten_kwargs) if whiten else self
        gates = _discover_gate_points(data, threshold, cluster_window)
```

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1606,98 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+            nzero = int(abs(tzero)*self.sample_rate.value)
+            npad = int(abs(tpad)*self.sample_rate.value)
+            half = nzero + npad
+            ntotal = 2*half
```
it looks like all of this is gate-independent, so could be moved outside the loop.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1606,98 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+                                           dt=self.dt, t0=self.t0)
+            window_timeseries[left_idx:right_idx] = window
+
+            out *= window_timeseries
```
Can you instead just do this and not have to create another timeseries?:

```python
            out[left_idx:right_idx] *= window
```

duncanmmacleod commented:
@tjma12, I have worked around the pycbc failure in #1053, please rebase.

tjma12 commented:
> @tjma12, I have worked around the pycbc failure in #1053, please rebase.

Rebased and squashed commits, I'll implement the other fixes once the build succeeds

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
 
+    def gate(self, tzero=1.0, tpad=0.5, whiten=True,
+             threshold=50., cluster_window=0.5, **whiten_kwargs):
+        """Discovers a set of indices where gating windows should be placed
```
I don't think this docstring is correct. Perhaps something like

```python
        """Mask clustered peaks above threshold using planck windows
 
        Parameters
        ------------
        ...
```

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+        Read data into a `TimeSeries`
+        >>> from gwpy.timeseries import TimeSeries
+        >>> data = TimeSeries.get('H1:GDS-CALIB_STRAIN', 1225542623,
+                                   1225542823)
```
Can you modify this example to use `fetch_open_data`?

tjma12 replied:
Done, switched this to use a large glitch from O1 open data

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+        >>> ax.plot(gated, label='Gated', color='orange', zorder=2)
+        >>> ax.set_xlim(1225542713, 1225542733)
+        >>> ax.legend()
+        >>> overlay.show()
```
Can you post in this ticket an example of the figure you get?

tjma12 replied:
See attached figure that is produced by updated docstring

![gating_example](https://user-images.githubusercontent.com/12533353/53436812-df9ac400-39b0-11e9-85bb-cd5c550b9295.png)

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+        """
+        # Find points to gate based on a threshold
+        data = self.whiten(**whiten_kwargs) if whiten else self
+        window_samples = cluster_window*data.sample_rate.value
```
```suggestion
        window_samples = cluster_window * data.sample_rate.value
```

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+        out = self.copy()
+
+        # Iterate over list of indices to gate and apply each one
+        nzero = int(abs(tzero)*self.sample_rate.value)
```
```suggestion
        nzero = int(abs(tzero) * self.sample_rate.value)
```

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+
+        # Iterate over list of indices to gate and apply each one
+        nzero = int(abs(tzero)*self.sample_rate.value)
+        npad = int(abs(tpad)*self.sample_rate.value)
```
```suggestion
        npad = int(abs(tpad) * self.sample_rate.value)
```

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1567,6 +1567,92 @@ def whiten(self, fftlength=None, overlap=0, method=DEFAULT_FFT_METHOD,
+        nzero = int(abs(tzero)*self.sample_rate.value)
+        npad = int(abs(tpad)*self.sample_rate.value)
+        half = nzero + npad
+        ntotal = 2*half
```
```suggestion
        ntotal = 2 * half
```

duncanmmacleod commented:
@tjma12, what's the status of this PR? It looks from @alurban's comment that you are going to put in one or more unit tests, which would be great.

tjma12 commented:
@duncanmmacleod I plan to address your requested changes and write up a unit test today

tjma12 commented:
I wrapped the data in an `abs()` when running the peak finding since whitened data can glitch negatively. Unit test is next.

duncanmmacleod commented:
@tjma12, it turns out that scipy.signal.find_peaks` is new in scipy-1.1.0, I don't think we should change it, but can you add some importerror protectionn so that users get told what is wrong when they try to run `TimeSeries.gate` and it doesn't work.

You can auto-skip the unit test using a decorator

```python
from ...testing.utils import skip_minimum_version

...

        @skip_minimum_version("scipy", "1.1.0")
        def test_gate(self, ...)
```

alurban commented:
@tjma12, what @duncanmmacleod said, I was going to suggest a direct import like this:

```python
try:
    from scipy.signal import find_peaks
except ImportError:
    raise ImportError('find_peaks is only available in scipy >= 1.1.0')
# Find points to gate based on a threshold
data = self.whiten(**whiten_kwargs) if whiten else self
window_samples = cluster_window * data.sample_rate.value
gates = find_peaks(abs(data.value), height=threshold,
                   distance=window_samples)[0]
```

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1626,6 +1626,12 @@ def gate(self, tzero=1.0, tpad=0.5, whiten=True,
+        try:
+             from scipy.signal import find_peaks
+        except ImportError:
+             raise ImportError("Must have scipy>=1.1.0 to utilize "
```
@tjma12, can you implement it a little different, to preserve the traceback:

```python
        except ImportError as exc:
            exc.args = ("scipy.signal.find_peaks is only available in scipy >= 1.1.0",)
            raise
````

tjma12 replied:
Both have been implemented
